src/nodes/dir.rs

```diff
@@ -262,6 +263,91 @@ impl Dir {
 
         Ok(state.attr)
     }
+
+    /// Gets the underlying path of the entry `name` in this directory.
+    ///
+    /// This also ensures that the entry is writable, which is determined by the directory itself
+    /// being mapped to an underlying path and the entry not being an explicit mapping.
+    fn get_writable_path(state: &mut MutableDir, name: &OsStr) -> NodeResult<PathBuf> {
+        if state.underlying_path.is_none() {
+            return Err(KernelError::from_errno(errno::Errno::EPERM));
+        }
+        let path = state.underlying_path.as_ref().unwrap().join(name);
+
+        if let Some(node) = state.children.get(name) {
+            if node.explicit_mapping {
+                return Err(KernelError::from_errno(errno::Errno::EPERM));
+            }
+        };
+
+        Ok(path)
+    }
+
+    // Same as `lookup` but with the node already locked.
+    fn lookup_locked(writable: bool, state: &mut MutableDir, name: &OsStr, ids: &IdGenerator,
+        cache: &Cache) -> NodeResult<(Arc<Node>, fuse::FileAttr)> {
+        if let Some(dirent) = state.children.get(name) {
+            let refreshed_attr = dirent.node.getattr()?;
+            return Ok((dirent.node.clone(), refreshed_attr))
+        }
+
+        let (child, attr) = {
+            let path = match state.underlying_path.as_ref() {
```

Comment:
Avoid using as_ref in this pattern -- it obscures what's going on.

Reply:
Done.